FEATURE: Add commands `media:listVariantPresets` and `media:removeVariants`

Jan3k3y opened the pull request:
Currently there is an option to create custom ImageVariant out of the `Settings.Neos.Media.yaml`. These VariantPresets can be edited, but variants with old/removed/outdated presets will not be deleted.

To solve this problem, I wrote a few lines of code and added two new commands...

1. list all found configurations for variants - from file and database, to get every single variant configuration, still existing or not
2. delete a variant preset by identifier and name provided by the other command

ISSUE: https://github.com/neos/neos-development-collection/issues/3086

**Upgrade instructions**

Nothing I know about so far

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [x] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

Jan3k3y committed:
FEATURE: imageVariant tidy up command

* finished command with better performance

ISSUE: https://github.com/neos/neos-development-collection/issues/3086

kitsunet commented on Neos.Media/Classes/Domain/Service/ImageVariantService.php:
```diff
@@ -0,0 +1,57 @@
+     * @param string $presetIdentifier
+     * @return array
+     */
+    public function getAllPresetsByConfigs(array $variantPresetConfigs, string $presetIdentifier = ''): array
```
Nitpicky, but we are commited to use as few shorts as possible, so I would prefer Configurations instead of Configs even if obivously in the context it's very clear what is meant here. Makes searching easier though.

And instead of the optional argument I would prefer a second method with two required parameters, so you can either (naming change suggetion) `getAllPresetsFromConfiguration` `getPresetFromConfigurationByName`

kitsunet commented on Neos.Media/Tests/Unit/Domain/Service/ImageVariantServiceTest.php:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * @return array
+     */
+    public function getAllPresetsByConfigsProvider(): array
```
Configs again

kitsunet commented on Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php:
```diff
@@ -0,0 +1,66 @@
+    /**
+     * Returns array of ImageVariants with outdated presets
+     *
+     * @param String[] $configuredPresets
```
You actually expect a dictionary/map here so this might be misleading, as the key is quite important and it cannot be a simple array. I think it would be good to use the `\Neos\Media\Domain\ValueObject\Configuration\Variant` and VariantPreset value objects to move those Presets around. Maybe add a PresetCollection even.

kitsunet commented on Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php:
```diff
@@ -0,0 +1,66 @@
+            ->from(ImageVariant::class, 'iv')
+            ->setMaxResults($limit);
+
+        if (!$isCliPresetFilter) {
```
This is unclear to me, maybe I need a coffee to understand the use of this, it seems to contradict the method name to not filter?

kitsunet reviewed (commented):
This is a cool addition, I left some nitpicks to go over, but it's on a good track. Thanks a lot!

Jan3k3y committed:
FEATURE: review implementation

* it is called configuration not configs!
* renamed ImageVariantService::getAllPresetsByConfigs to getAllPresetsOfIdentifier
* refactored ImageVariantService::getAllPresetsOfIdentifier for usability
* specified configuresPresets type in ImageVariantRepository::findAllWithOutdatedPresets
* better naming for query condition + example

ISSUE: https://github.com/neos/neos-development-collection/issues/3086

Jan3k3y commented:
I have adjusted some things and hope that it is now more understandable and fits to your naming commitment.

Jan3k3y commented:
When I re-requested the review from kitsunet, Sebobo was unintentionally removed from the reviewers....
But I can not add him again

Jan3k3y commented:
Does anyone have time for another review?

ahaeslich commented:
@kitsunet Do we want to include this in the 8.3 release?

kitsunet commented on Neos.Media/Classes/Command/MediaCommandController.php:
```diff
@@ -276,7 +297,7 @@ public function removeUnusedCommand(string $assetSource = '', bool $quiet = fals
         if ($assumeYes === false) {
             if (!$this->output->askConfirmation(sprintf('Do you want to remove <b>%s</b> unused assets?', $unusedAssetCount))) {
-                exit(1);
+                exit;
```
```suggestion
                exit(1);
```

I think this should stay, it is an unsuccessful run if nothing was deleted after all?

bwaidelich replied:
why not `$this->quit(1);`?

kitsunet replied:
or that, sure!

kitsunet commented on Neos.Media/Classes/Command/MediaCommandController.php:
```diff
@@ -388,6 +409,161 @@ public function renderThumbnailsCommand(int $limit = null, bool $quiet = false)
+            }
+            if (!$this->output->askConfirmation(PHP_EOL . 'Do you want to remove all variants with outdated presets? [Y,n] ')) {
+                $this->output->outputLine(PHP_EOL . '<em>No variants have been deleted...</em>');
+                exit;
```
```suggestion
                exit(1);
```

Same as above.

kitsunet commented on Neos.Media/Classes/Command/MediaCommandController.php:
```diff
@@ -388,6 +409,161 @@ public function renderThumbnailsCommand(int $limit = null, bool $quiet = false)
+            } catch (IllegalObjectTypeException $e) {
+                $this->output->outputLine(PHP_EOL . 'Unable to remove %s: "%s"', [get_class($variantToRemove), $variantToRemove->getTitle()]);
+                $this->output->outputLine(PHP_EOL . $e->getMessage());
+                exit;
```
```suggestion
                exit(1);
```

Same as above.

kitsunet commented on Neos.Media/Classes/Command/MediaCommandController.php:
```diff
@@ -388,6 +409,161 @@ public function renderThumbnailsCommand(int $limit = null, bool $quiet = false)
+
+        if (empty($currentPresets)) {
+            !$quiet && $this->output->outputLine(PHP_EOL . PHP_EOL . '<em>No presets found.</em>');
+            exit;
```
```suggestion
            exit(1);
```

Same as above. Maybe debatable here and I am fine either way as the user cannot know if presets exist or not, so this might be a valid early exit condition.

kitsunet commented on Neos.Media/Classes/Command/MediaCommandController.php:
```diff
@@ -388,6 +409,161 @@ public function renderThumbnailsCommand(int $limit = null, bool $quiet = false)
+
+        if (empty($outdatedVariants)) {
+            !$quiet && $this->output->outputLine(PHP_EOL . PHP_EOL . '<em>No outdated variants found.</em>');
+            exit;
```
```suggestion
            exit(1);
```

Same as above.

kitsunet commented on Neos.Media/Classes/Command/MediaCommandController.php:
```diff
@@ -388,6 +409,161 @@ public function renderThumbnailsCommand(int $limit = null, bool $quiet = false)
+        $variants = $this->imageVariantRepository->findAllWithOutdatedPresets($currentPresets, !empty($presetIdentifier), $filterCroppedVariants, $limit);
+        if (empty($variants)) {
+            !$quiet && $this->output->outputLine(PHP_EOL . PHP_EOL . '<em>No variants found.</em>');
+            exit;
```
```suggestion
            exit(1);
```

Same as above.

kitsunet reviewed (approved):
Minor change requests/questions. But wouldn't utlimately block merging.

Jan3k3y commented:
Should I change it directly in this MR or create a new one and change it in `removeUnusedCommand` too?

crydotsnake commented:
> Should I change it directly in this MR or create a new one and change it in `removeUnusedCommand` too?

If you mean the suggestions from @kitsunet you can change it in this PR.

crydotsnake commented:
We should also not forget, to add the new command to the Command Reference in the documentation:

https://github.com/neos/neos-development-collection/blob/8.3/Neos.Neos/Documentation/References/CommandReference.rst#package-neosmedia

Jan3k3y committed:
FEATURE: reworked exit condition

* changed from 'exit' to '$this->quit(1)'
* slightly edited phpDoc

ISSUE: https://github.com/neos/neos-development-collection/issues/3086

Jan3k3y committed:
FEATURE: reworked exit condition

* changed from 'exit' to '$this->quit(1)' in removeUnusedCommand

ISSUE: https://github.com/neos/neos-development-collection/issues/3086

Jan3k3y committed:
FEATURE: refactoring with marc henry

- count() instead of sizeof()
- new __construct in ImageVariantService
- test: dataProvider rework with yield and less code

Jan3k3y committed:
FEATURE: review implementation from neos sprint

- seperated monster command into two smaller ones
 (1) list all existing variant preset configurations
 (2) remove all variants with provided identifier and name

mficzel reviewed (dismissed):
Just to document the discusssion on the neos sprint:

The complex command choud be split into two commands which would make it more versatile and easy to review. 

One command would allow to show stats about existing variantPresets with an option to highlight/filter those that exist but are not in the configuration. 

The second command would clear the variants of the given preset.

mficzel commented on Neos.Media/Classes/Command/MediaCommandController.php:
```diff
@@ -388,6 +410,164 @@ public function renderThumbnailsCommand(int $limit = null, bool $quiet = false)
+            $this->outputLine('To delete them, please use the <b>removeUnusedImageVariants</b> command.');
+        }
+
+        die;
```
why `die`?

Jan3k3y replied:
That is a pretty good question 😅

mficzel commented on Neos.Media/Classes/Command/MediaCommandController.php:
```diff
@@ -388,6 +410,164 @@ public function renderThumbnailsCommand(int $limit = null, bool $quiet = false)
+     *
+     * @throws StopCommandException
+     */
+    public function removeOutdatedVariantsCommand(string $identifier, string $variantName, bool $quiet = false, bool $assumeYes = false, int $limit = null)
```
Would it make sense to have a distinct `removeCustomCropVariantsCommand` instead of setting id and name to `NULL`.

ahaeslich replied:
Yes, I would also split this in a distinct command as `removeoutdatedvariants` implies that a subset of variants is cleaned. Using `--identifier NULL --variant-name NULL` would AFAIK delete *all* custom crops.

That sounds kind of harmful for me.

Additionally we should split this use case in another issue and not add it on top of this changes.

Jan3k3y replied:
Removed, but saved for a separat command 😉

ahaeslich replied:
Addition: Is the function name still correct? You can now remove variants that are still configured.

mficzel reviewed (commented):
I generally like this now :-) cannot really test as i have no case for variants here. 

btw. What will happen if i remove a custom crop that is used.

Jan3k3y committed:
FEATURE: review implementations

- removed 'die'
- excluded option to delete custom crops

Jan3k3y commented:
@mficzel 
> btw. What will happen if i remove a custom crop that is used.

The short form: 
It will be removed

The long form: 
If you remove a not configured variant this case is not possible (you can only use configured variants).
If you remove a configured variant there will be an error at the location it would be rendered (don't know the error message but there will be one). To fix this you have to regenerate them with `./flow media:renderVariants`.

Do you think I should ask/inform the user in case a configured variant will be deleted or give the advise to use the `./flow media:renderVariants` command after execution?

ahaeslich commented on Neos.Media/Classes/Command/MediaCommandController.php:
```diff
@@ -388,6 +410,136 @@ public function renderThumbnailsCommand(int $limit = null, bool $quiet = false)
+                    $outputPresets[] = [$presetIdentifier, $key, $presetVariantNames[$key]];
+                }
+            }
+            $this->outputLine('There is a <b>configuration</b> with following <b>variant presets:</b>');
```
```suggestion
            $this->outputLine('There is <b>no configuration</b> for the following <b>variant presets:</b>');
```

ahaeslich commented on Neos.Media/Classes/Command/MediaCommandController.php:
```diff
@@ -388,6 +410,136 @@ public function renderThumbnailsCommand(int $limit = null, bool $quiet = false)
+                }
+            }
+            $this->outputLine('There is a <b>configuration</b> with following <b>variant presets:</b>');
+            $this->output->outputTable($outputPresets, ['Configured Presets', 'Label of variant']);
```
I would add `identifier` and `variantName` to the table headers as they are needed in the remove command and it is not obvious which is which. (Also relevant on line 459)

ahaeslich commented on Neos.Media/Classes/Command/MediaCommandController.php:
```diff
@@ -388,6 +410,136 @@ public function renderThumbnailsCommand(int $limit = null, bool $quiet = false)
+    }
+
+    /**
+     * Tidy up outdated imageVariants with provided identifier and variant name.
```
This description is not correct anymore as you can also delete configured variants.

Jan3k3y committed:
FEATURE: review implementations

- reworked namings, UI messages and description
- refactored usage of identifier and variantName
- added info what to do if removed configured preset

mhsdesign commented:
Hi ;) what is the status on this? It seems some things changed for the better after we reviewed it at the sprint (eg. we have two commands instead of one big command - i do really like this way ^^)

are there any leftovers to be discussed? or any critical things like: upsi this command deletes wayyyyyy to much?

ahaeslich committed:
TASK: Inlined ImageVariantService and removed dead code

- Fixed output of outdated variant
- Optimized description and output texts
- Removed test of unused ImageVariantService

Thanks @mhsdesign for the pair review

ahaeslich committed:
TASK: Rename `variantLabel` to `variantName`

mhsdesign reviewed (approved):
reviewed, tested and refactored with @ahaeslich bitteschön

mhsdesign commented:
to get tests working, it could be rebased

mhsdesign commented on Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php:
```diff
@@ -0,0 +1,96 @@
+ */
+class ImageVariantRepository extends AssetRepository
+{
+    public function findVariantsByName(?string $identifier, ?string $variantName, int $limit = null): array
```
hold my beer, why are the parameters ?nullable

mhsdesign replied:
its not used with nullable params, so we can remove the nullablity and also below: `if (is_null($identifier) && is_null($variantName)) {`

mhsdesign commented on Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php:
```diff
@@ -0,0 +1,96 @@
+     * @param int|null $limit
+     * @return ImageVariant[]
+     */
+    public function findAllWithOutdatedPresets(array $configuredPresets, int $limit = null): array
```
i think $configuredPresets should not be a dependency as parameter but rather 

```
$currentPresets = $this->assetVariantGenerator->getVariantPresets();
```

should be used

mhsdesign reviewed (changes requested):
send @ahaeslich a patch ;)

ahaeslich committed:
TASK: Refactor `ImageVariantRepository`

- Remove nullable parameters
- `getVariantPresets` directly instead of applying it as function parameter

Thanks @mhsdesign for the patch

ahaeslich commented:
> send @ahaeslich a patch ;)

Thx for the patch ❤️ Checked and pushed it
